app/src/forms/file/storage/objectStorageService.js

```diff
@@ -0,0 +1,213 @@
+          if (err) {
+            // doesn't throw a 404 when given a bad key
+            reject(err);
+          } else {
+            resolve(data);
+          }
+        });
+      });
+    } catch (e) {
+      errorToProblem(SERVICE, e);
+    }
+  }
+
+  async read(fileStorage) {
+    try {
+      const params = {
+        Bucket: this._bucket,
+        Key: fileStorage.path
+      };
+      return new Promise((resolve, reject) => {
+        const _local_s3 = this._s3;
+        // eslint-disable-next-line
+        _local_s3.headObject(params, function (err, data) {
+          if (err) {
+            if (404 === err.statusCode) {
+              reject(new Problem(404, 'File not found'));
+            } else {
+              reject(err);
+            }
+          } else {
+            // want to return the stream...
+            resolve(_local_s3.getObject(params).createReadStream());
+          }
+        });
+      });
+    } catch (e) {
+      errorToProblem(SERVICE, e);
+    }
+  }
+
+  async move(fileStorage, ...subdirs) {
+    try {
+      const sourcePath = fileStorage.path;
+      const file = await this.copyFile(fileStorage, ...subdirs);
+      if (file) {
+        // this doesn't return the new key/path, but we can build it
+        const newPath = this._join(this._key, ...subdirs, fileStorage.id);
+        // now delete original...
+        const params = {
+          Bucket: this._bucket,
+          Key: sourcePath
+        };
+
+        return new Promise((resolve, reject) => {
+          // eslint-disable-next-line no-unused-vars
+          this._s3.deleteObject(params, (err, data) => {
+            if (err) {
+              reject(err);
+            } else {
+              resolve(newPath);
+            }
+          });
+        });
+      }
+    } catch (e) {
+      errorToProblem(SERVICE, e);
+    }
+  }
+
+  async copyFile(fileStorage, ...subdirs) {
+    try {
+      const destPath = this._join(...subdirs);
+
+      const params = {
+        Bucket: `${this._bucket}/${this._key}${destPath}`,
+        CopySource: `${this._bucket}/${fileStorage.path}`,
+        Key: fileStorage.id
+      };
+
+      return new Promise((resolve, reject) => {
+        // eslint-disable-next-line no-unused-vars
+        this._s3.copyObject(params, (err, data) => {
+          if (err) {
+            reject(err);
+          } else {
+            resolve(data);
+          }
+        });
+      });
+    } catch (e) {
+      errorToProblem(SERVICE, e);
+    }
+  }
+
+}
+
+const endpoint = config.get('files.objectStorage.endpoint');
+const bucket = config.get('files.objectStorage.bucket');
+const key = config.get('files.objectStorage.key');
+const accessKeyId = config.get('files.objectStorage.accessKeyId');
+const secretAccessKey = config.get('files.objectStorage.secretAccessKey');
+
+let objectStorageService = new ObjectStorageService({accessKeyId: accessKeyId, secretAccessKey: secretAccessKey, endpoint: endpoint, bucket: bucket, key: key});
```

Comment:
As an aside, you could potentially get away with just getting `config.get('files.objectStorage')` and then pass that object directly to the ObjectStorageService constructor (all the object parameters are named the same way already).

Also, more async and file sync considerations as with earlier comment.

Reply:
could do, but prefer to be explicit and have it blow up earlier in the call stack if values are not in config.  line error will point out which one is failing.
async is more for the storageService interface, if we ever put in storage that only has async calls, we don't have to change the interface at all. if we see speed issues with any sync calls in *StorageService, we can flip to async (if supported) without changing storageService or it's callers.